### PR TITLE
Simplification

### PR DESCRIPTION
**Pull request by dkfellows**
This branch substantially simplifies the DSG core, mostly by factoring out functionality so that range and type-range checks are done in exactly the same (correct!) way everywhere.

### CODE REVIEW

**Comment by dkfellows** [comment]
Return to this after #37 has gone in; this will need conflict resolution.

**Comment by rowleya** [comment]
I tried to do some merging here but there are some non-obvious bits that will depend on what you were doing here, so I'll leave it to you!

**Comment by dkfellows** [comment]
This isn't fully baked yet.

**Comment by dkfellows** [comment]
TODO: document the variables in the DSG code better

**Comment by alan-stokes** [comment]
looks good to me. jsut fix merge issues then all good to go

**Comment by dkfellows** [comment]
OK, my changes particularly allow the transformation of something like this:

        cmd_word = (
            (cmd_data_len << 28) |
            (Commands.WRITE.value << 20) |
            (repeat_reg_usage << 16) | (data_len << 12) | parameters)
        padding = 4 - data_type.size if data_type.size < 4 else 0
        cmd_word_list = struct.pack(	
            "<I{}{}x".format(data_type.struct_encoding, padding),	
            cmd_word, _rescale(data, data_type))

into this equivalent (but hugely more comprehensible) form:

        cmd_word = _binencode(Commands.WRITE, {
            _Field.LENGTH: cmd_data_len,
            _Field.USAGE: repeat_reg_usage,
            _Field.DESTINATION: data_len,
            _Field.IMMEDIATE: parameters})
        data_word = data_type.encode(data)
        cmd_word_list = cmd_word + data_word

File: data_specification/data_specification_executor_functions.py [hunk 1]
```diff
@@ -367,17 +352,27 @@ def _write_bytes_to_mem(self, data, command):
...
             raise RegionNotAllocatedException(self.current_region, command)
 
+        self.__write_blob(data)
+
+    def __write_blob(self, data):
+        """ Does the actual write to the region, enforcing that writes cannot\
+            go outside the region.
+
+        :param blob: The data to write
+        :type blob: str
```
Comment (rowleya) [inline data_specification/data_specification_executor_functions.py]: The parameter isn't actually called blob

Comment (rowleya) [reply data_specification/data_specification_executor_functions.py]: But I fixed it :)

**Pull request merged**